delta_backend/src/extractor.py

```diff
@@ -0,0 +1,409 @@
+        """
+        if not date or not isinstance(date, str):
+            self._log_error(field_name, date, "Invalid value. Must be non empty string")
+            return ""
+        try:
+            dt = datetime.fromisoformat(date)
+            return dt.strftime(format)
+        except ValueError as e:
+            self._log_error(field_name, date, e)
+            return ""
+    
+    def _convert_date_to_safe_format(self, field_name, date) -> str:
+        try:
+            dt = datetime.fromisoformat(date)
+            if dt.tzinfo is None:
+                dt = dt.replace(tzinfo=timezone.utc)
+        except Exception as e:
+            self._log_error(field_name, date, e)
+            return ""
+
+        # Allow only +00:00 or +01:00 offsets (UTC and BST) and reject unsupported timezones
+        offset = dt.utcoffset()
+        allowed_offsets = [timedelta(hours=0), timedelta(hours=1)]
+        if offset is not None and offset not in allowed_offsets:
+            self._log_error(field_name, date, "Unsupported Format or offset")
+            return ""
+
+        # remove microseconds
+        dt_format = dt.replace(microsecond=0)
+
+        formatted = dt_format.strftime("%Y%m%dT%H%M%S%z")
+        return formatted.replace("+0000", "00").replace("+0100", "01")
+    
+    def extract_nhs_number(self):
+        patient = self._get_patient()
+        if patient:
+            identifier_list = patient.get("identifier", [])
+            for identifier in identifier_list:
+                if identifier.get("system", "") == self.NHS_NUMBER_SYSTEM_URL:
+                    return identifier.get("value", "")
+        return ""
+            
+    def extract_person_forename(self):
+        return self._get_person_names()[0]
+    
+    def extract_person_surname(self):
+        return self._get_person_names()[1]
+    
+    def extract_person_dob(self):
+        patient = self._get_patient()
+        
+        if patient:
+            dob = patient.get("birthDate", "")   
+            
+            return self._convert_date("PERSON_DOB", dob, self.DATE_CONVERT_FORMAT)          
+        return ""
+    
+    def extract_person_gender(self):
+        patient = self._get_patient()
+        if patient: 
+            gender = patient.get("gender", "").upper()
+            try:
+                return Gender[gender].value
+            except KeyError:
+                return ""   
+        return ""
+    
+    def extract_valid_address(self):
+        occurrence_time = self._get_occurance_date_time()
+        patient = self._get_patient()
+        
+        addresses = patient.get("address", [])
+        if not isinstance(addresses, list) or not addresses:
+            return self.DEFAULT_POSTCODE
+
+        valid_addresses = [a for a in addresses if "postalCode" in a and self._is_current_period(a, occurrence_time)]
+        if not valid_addresses:
+            return self.DEFAULT_POSTCODE
```

Comment:
If there's only one address, the spec says to use it even if it's not current at the time of vaccination - follow up ticket?